### DISCUSSION TITLE
Barcodes that are both UPC-A and EAN

### DISCUSSION THREAD

**Discussion by duanqn** [zxing-cpp/zxing-cpp#419] [Q&A] [answered]
I have noticed that some barcodes are both valid in UPC-A and EAN, but their decoded digits are different.
Does ZXing-cpp only return one standard in this situation?
Is it possible to post-process ZXing-cpp results to pick out these codes and generate the other alternative?

**Comment by oehhar** [comment]
I suppose, the decoded digits diffence in a leading 0. It is good practice to fill-up left any EAN/UPC/DataBar/GS1 AI 01 code to 14 digits to be comparable. The same item is identified by all of those codes, as they all rely on the same GTIN.
If it does not decode on the same number (when filled left to 14 digits), it is IMHO a bug.

Take care,
Harald

**Reply by duanqn** [reply]
Oh right, they only differ in leading zeros. Does that mean any 12-digit UPC code can be zero-expanded to EAN?

**Comment by oehhar** [comment, marked as answer]
Yes.
EAN can encode 13 digits, UPC-A 12. As EAN is a much weaker encoding (higher error probability), it is advisable, to encode 12 digits as UPC and 13 digits as EAN. All readers have "emulation modes" and output directly wether "no leading zeros" or always 13 or even 14 digits.

**Comment by axxel** [comment]
zxing-cpp currently still follows ZXing's (original Java library) behavior of 'preferring' UPC-A over EAN13, i.e. when both codes are enabled and a valid UPC-A symbol is found, it is reported back as a 12 digit UPC-A symbol. Ever since I touched that code I had the feeling that reporting it back as 13-digit EAN-13 would actually be the better option, but feared that this might just be my "european" vs. "US" viewpoint.

I'd be happy to change that behavior if there is something more substantial than my own gut feeling supporting it. :)

**Comment by oehhar** [comment]
I think, it is ok as it is.
Note: "EAN: European Article Numbering" is the 13 digit extension of UPC. Nowdays, it has nothing to do with Europe. It was only created and called like that to support other countries then USA. The US term "ROW" (for "Rest of World") would be more appropriate ;-).

**Comment by duanqn** [comment]
Got it. Thanks everyone!
